Supports building under libvirt provider

conorsch opened the pull request:
The previous config assumes that VirtualBox is used as the Vagrant
backend, and accordingly bumped the specs (RAM and CPU count) only under
VirtualBox. Although the previous config still created the `grsec-build`
host under the vagrant-libvirt backend when explicitly specified, the
host was created with only 512MB of RAM, which is insufficient for
building a kernel.

The existing logic for dynamically determining the max CPU count on the
host machine is now applied to libvirt as well as virtualbox for the
backend provider. Also bumping RAM to 2048MB, same as under virtualbox.

conorsch committed:
Supports building under libvirt provider

The previous config assumes that VirtualBox is used as the Vagrant
backend, and accordingly bumped the specs (RAM and CPU count) only under
VirtualBox. Although the previous config still created the `grsec-build`
host under the vagrant-libvirt backend when explicitly specified, the
host was created with only 512MB of RAM, which is insufficient for
building a kernel.

The existing logic for dynamically determining the max CPU count on the
host machine is now applied to libvirt as well as virtualbox for the
backend provider. Also bumping RAM to 2048MB, same as under virtualbox.

psivesely commented:
1. Install vagrant libvirt provider via instructions on https://github.com/vagrant-libvirt/vagrant-libvirt
2. Install vagrant mutate provider via https://github.com/sciurus/vagrant-mutate
3. Mutate Virtualbox Vagrant image to support libvirt
4. Fails with the following error:
```
% vagrant up grsec-build --provider libvirt                                                                                             Bringing machine 'grsec-build' up with 'libvirt' provider...
Error while connecting to libvirt: Error making a connection to libvirt URI qemu:///system?no_verify=1&keyfile=/home/fowlslegs/.ssh/id_rsa:
Call to virConnectOpen failed: Failed to connect socket to '/var/run/libvirt/libvirt-sock': No such file or directory
```

You mentioned you found a community-maintained libvirt role that should set things up correctly to resolve this issue. Will retry after we get that setup on my machine.

psivesely commented:
So I got this working finally after building a new kernel and running the `libvirt` playbook, both versions from the `vagrant-kvm` branch of infra. I also had to make one change to the Vagrantfile, specifically to make sure the guest CPU configuration was such that it was not requesting the non-existent svm (AMD-V) feature of my processor. This change was suggested in https://github.com/vagrant-libvirt/vagrant-libvirt/issues/667#issuecomment-256897999.

I think you should read the section https://libvirt.org/formatdomain.html#elementsCPU. `host-passthrough`, which the commit I'm pushing here uses, in practice seems unlikely to ever be the culprit in some passed over or host-specific bug, but the docs do warn:

> the downside of this mode is that the guest environment cannot be reproduced on different hardware.

An alternative to consider is specifying a particular processor to emulate (viz., the ones on the NUCs we're recommending) with `custom` for "ultimate" environment reproducibility.

As to the libvirt bug itself, the documentation does give some warnings on the unreliability of `host-model`, (which is set by default in the config file Vagrant autogenerates), and a little more to that regard can be found in https://www.redhat.com/archives/libvir-list/2016-May/msg01940.html.

Commit:
Use host-passthrough for Vagrant libvirt CPU config

The libvirt documentation warns:

> Beware, due to the way libvirt detects host CPU and due to the fact libvirt
> does not talk to QEMU/KVM when creating the CPU model, CPU configuration
> created using host-model may not work as expected. The guest CPU may differ
> from the configuration and it may also confuse guest OS by using a combination
> of CPU features and other parameters (such as CPUID level) that don't work.
> Until these issues are fixed, it's a good idea to avoid using host-model and
> use custom mode with just the CPU model from host capabilities XML.

This was causing errors on my workstation, but not Conor's, which could be due
to differences in libvirt versions or hardware. Whatever the cause, the
documentation makes clear this setting may be error-prone as bugs are worked
out, so it seems most expedient to set `cpu_mode` to `host-passthrough` as
suggested in
https://github.com/vagrant-libvirt/vagrant-libvirt/issues/667#issuecomment-256897999
until libvirt works out it's bugs or we decide to switch to a custom config as
suggested in
https://github.com/freedomofpress/ansible-role-grsecurity/pull/85#issuecomment-266611369.

conorsch commented:
Hmm, interesting. Good research on this one! Can you set the passthrough option in your `~/.vagrant.d/Vagrantfile` and have it get merged correctly? This this seems to be host-specific, I'd prefer to keep the config option addition host-specific, and then document it commented-out here.

Commit:
Comment out host-passthrough CPU for libvirt provider

Leaving this here with the note in case others run into the problem. If they
look further into the issue by following the link or are already familiar with
Vagrant they'll figure out they can do this globally as I now am.

conorsch commented:
Great addition, @fowlslegs! Goin' in.
